src/lib/shim/shadow_sem.c

```diff
@@ -0,0 +1,158 @@
+
+#include "lib/logger/logger.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <linux/futex.h>
+#include <semaphore.h>
+#include <stdalign.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+
+// These could just be at global scope, but calling this at the specific points
+// where we use  the futex is a bit less fragile.
+static inline void _value_is_valid_futex(const shadow_sem_t* s) {
+    static_assert(sizeof(s->_value) == 4, "futex must be exactly 4 bytes large");
+    static_assert(alignof(typeof(s->_value)) >= 4, "futex must be >= 4 aligned");
+}
+
+// Perform a wake operation on the futex in `s`.
+static int _futex_wake(const shadow_sem_t* s) {
+    _value_is_valid_futex(s);
+    return (int)syscall(SYS_futex, &s->_value, FUTEX_WAKE, 1, NULL, NULL, 0);
+}
+
+// Perform a wait operation on the futex in `s`, with an absolute timeout.
+static int _futex_wait_abs(shadow_sem_t* s, const struct timespec* timeout) {
+    _value_is_valid_futex(s);
+    // Unlike FUTEX_WAIT, FUTEX_WAIT_BITSET uses an absolute timeout.
+    return (int)syscall(
+        SYS_futex, &s->_value, FUTEX_WAIT_BITSET, 0, timeout, NULL, FUTEX_BITSET_MATCH_ANY);
+}
+
+int shadow_sem_init(shadow_sem_t* sem, int pshared, unsigned int _value) {
+    if (_value > SEM_VALUE_MAX) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    *sem = (shadow_sem_t){
+        ._value = ATOMIC_VAR_INIT(_value),
+        ._nwaiters = ATOMIC_VAR_INIT(0),
+    };
+
+    return 0;
+}
+
+int shadow_sem_destroy(shadow_sem_t* sem) {
+    // Nothing to do.
+    return 0;
+}
+
+int shadow_sem_post(shadow_sem_t* sem) {
+    uint32_t prev_value = atomic_load_explicit(&sem->_value, memory_order_relaxed);
+    while (1) {
+        if (prev_value >= SEM_VALUE_MAX) {
+            errno = EOVERFLOW;
+            return -1;
+        }
+        // We use memory_order_seq_cst to get a global total ordering of the
+        // operations on `_nwaiters` together with the operations on `_value`.
+        if (atomic_compare_exchange_weak_explicit(&sem->_value, &prev_value, prev_value + 1,
+                                                  memory_order_seq_cst, memory_order_relaxed)) {
+            break;
+        }
+    }
+
+    // If we didn't see a futex value of 0, we never need to do a wakeup.  A
+    // concurrent thread that's trying to wait on the semaphore can't end up
+    // sleeping on a non-zero _value, as enforced by the futex operation.
+    if (prev_value != 0) {
+        return 0;
+    }
+
+    // If no threads are asleep on the futex, we don't need to do a wakeup
+    // operation.  While there is some cost and complexity for tracking
+    // `_nwaiters`, this gives about a 5% performance improvement in the phold
+    // mezzo benchmark.
+    uint32_t nwaiters = atomic_load_explicit(&sem->_nwaiters, memory_order_seq_cst);
+    if (nwaiters == 0) {
+        return 0;
+    }
+
+    long futex_rv = _futex_wake(sem);
```

Comment:
Shouldn't this be `int futex_rv` to match the return type of _futex_wake?

Reply:
Done